governance/program/src/state.rs

```diff
@@ -0,0 +1,247 @@
+//! Program accounts
+
+use solana_program::pubkey::Pubkey;
+
+/// Max length of a governance name
+pub const MAX_GOVERNANCE_NAME_LENGTH: usize = 32;
+
+/// Max length of a proposal description link
+pub const MAX_PROPOSAL_DESCRIPTION_LINK_LENGTH: usize = 200;
+
+/// Max length of a proposal name
+pub const MAX_PROPOSAL_NAME_LENGTH: usize = 32;
+
+/// Max length of a proposal's instruction data
+pub const MAX_PROPOSAL_INSTRUCTION_DATA_LENGTH: usize = 450;
+
+/// Max number of transactions allowed for a proposal
+pub const MAX_TRANSACTIONS: usize = 5;
+
+/// Defines all Governance accounts types
+#[derive(Clone, Debug, PartialEq)]
+pub enum GovernanceAccountType {
+    /// 0 - Default uninitialized account state
+    Uninitialized,
+
+    /// 1 - Governance account
+    Governance,
+
+    /// 2 - Proposal account for Governance account. A single Governance account can have multiple Proposal accounts
+    Proposal,
+
+    /// 3 - Proposal voting state account. Every Proposal account has exactly one ProposalState account
+    ProposalState,
+
+    /// 4 - Vote record account for a given Proposal.  Proposal can have 0..n voting records
+    VoteRecord,
+
+    /// 5 Custom Single Signer Transaction account which holds instructions to execute for Proposal
+    CustomSingleSignerTransaction,
+}
+
+impl Default for GovernanceAccountType {
+    fn default() -> Self {
+        GovernanceAccountType::Uninitialized
+    }
+}
+
+/// Vote  with number of votes
+#[derive(Clone, Debug, PartialEq)]
+pub enum Vote {
+    /// Yes vote
+    Yes(u64),
+
+    /// No vote
+    No(u64),
+}
+
+/// Governance Account
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Governance {
+    /// Account type
+    pub account_type: GovernanceAccountType,
+
+    /// Voting threshold in % required to tip the vote
+    pub vote_threshold: u8,
+
+    /// Minimum slot time-distance from creation of proposal for an instruction to be placed
+    pub minimum_slot_waiting_period: u64,
+
+    /// Governance mint
+    pub governance_mint: Pubkey,
+
+    /// Council mint
+    pub council_mint: Option<Pubkey>,
+
+    /// Program ID that is governed by this Governance
+    pub program: Pubkey,
+
+    /// Time limit in slots for proposal to be open to voting
+    pub time_limit: u64,
```

Comment:
What does this decide? If we don't sign fast enough, the proposal goes into `Expired` before ever going into `Voting`?

Reply:
It's the max time allocated for voting. I renamed it to `max_voting_time` to be self explanatory.